driver: i2c: TCA9546a: Add support for more I2C mux

aaillet opened the pull request:
Hi, 
Initial goal was to support the the PCA9548 8 channel mux I'm working with.
As discussed with @kurddt , this PR is containing modification in order for the I2C mux driver to :
- Support more mux from the TCA954x and PCA954x families.
- Support any mux size that could exist in these families.
- Support reset signals for mux that are equipped with a reset input pin.

This driver has been tested with an 8-channel with reset PCA9548A mux from TI and should not break compatibility with previously tested mux.

kurddt reviewed (changes requested):
I don't know what is the policy about changing existing driver's compatible property. @mbolivar could you advice on the matter ?

aaillet commented:
Rebased on top of main !

kurddt commented:
After some looking around in the zephyr source tree and thoughts, I think it would be better from the driver user's point of view to handle the different variant of the i2c mux like what is done for the mcp320X or lmp90xxx adc drivers:

A binding file with different compatible exists for each supported variant. That way the `nchans` devicetree property is not needed but deduced from the compatible. 

That said I won't block the PR for that, if you don't want to make that change I will do this work after the PR is merged

aaillet commented:
I don't know if it's the right solution to multiply the number of binding (one per reference) especially since they are for the most part, empty.
But if @mbolivar confirms that this is the solution applied in our case, we might as well not enter this PR and do it directly to avoid breaking the "compatible" several times.
In this case, I could do it myself based on the drivers you pointed out.

mbolivar-nordic commented:
Changing compatibles is not ideal, but we do do it sometimes if we have to.

Can you please update the release notes to reflect what happened and how users can recover, though? Here is an example from 2.7:

https://docs.zephyrproject.org/latest/releases/release-notes-2.7.html#build-and-infrastructure

(See "Devicetree bindings" section in that link)

kurddt commented:
@mbolivar-nordic and what are your thoughts on having several compatibles ? one for each supported device

mbolivar-nordic commented:
@kurddt in general? I am in favor of it wherever it makes sense. Note that you can create multiple bindings that share properties using `include`, so e.g.

```
# my-common-binding.yaml
properties:
  foo:
     ...
  bar:
     ...

# vnd,device1.yaml
compatible: vnd,device1
include: my-common-binding.yaml

# vnd,device2.yaml
compatible: vnd,device2
include: my-common-binding.yaml
```

etc.

If that makes sense here, then I'm in favor of doing that.

```

kurddt commented:
@aaillet Could you make the change ? and sorry to not have thought of that earlier

aaillet commented:
> @aaillet Could you make the change ? and sorry to not have thought of that earlier

Hi, I'm gonna do the changes soon.

aaillet commented:
Hi, just pushed new version of the driver.
The driver is now managing multiple compatible fields (tca9546a and tca9548a at the moment).
Each device version binding file is now declaring its own device compatible field.
Each device binding is also declaring its own child binding compatible to ensure backward compatibility and simple dts fields completion for users.
The driver has been tested using a pca9548a from NXP and everything is working as expected.
if I forgot to mention something, don't hesitate.
Aymeric

aaillet commented:
Please @mbolivar-nordic , tell me if this bindings/device tree implementation is acceptable. Thanks

kaolpr commented:
I've tried this PR on TCA9548A and can confirm that it works. However, I had to change `I2C_TCA954X_CHANNEL_INIT_PRIO` to be of value greater than `I2C_TCA954X_CHANNEL_ROOT_PRIO` to avoid not ready error (to initialize root prior to channels).

aaillet commented:
> I've tried this PR on TCA9548A and can confirm that it works. However, I had to change `I2C_TCA954X_CHANNEL_INIT_PRIO` to be of value greater than `I2C_TCA954X_CHANNEL_ROOT_PRIO` to avoid not ready error (to initialize root prior to channels).

Hi, thank you for testing.
When testing on my board, I also have to set `I2C_TCA954X_CHANNEL_INIT_PRIO` to a greater value than `I2C_TCA954X_CHANNEL_ROOT_PRIO` following an advice that @kurddt gave [this issue](https://github.com/zephyrproject-rtos/zephyr/issues/37786#issuecomment-901862285).
Maybe we should set these by default for any board, but I let @kurddt tell us if he thinks it's a good idea !

kurddt commented:
> > I've tried this PR on TCA9548A and can confirm that it works. However, I had to change `I2C_TCA954X_CHANNEL_INIT_PRIO` to be of value greater than `I2C_TCA954X_CHANNEL_ROOT_PRIO` to avoid not ready error (to initialize root prior to channels).
> 
> Hi, thank you for testing. When testing on my board, I also have to set `I2C_TCA954X_CHANNEL_INIT_PRIO` to a greater value than `I2C_TCA954X_CHANNEL_ROOT_PRIO` following an advice that @kurddt gave in [this issue](https://github.com/zephyrproject-rtos/zephyr/issues/37786#issuecomment-901862285). Maybe we should set these by default for any board, but I let @kurddt tell us if he thinks it's a good idea !

The best solution would be that I2C_TCA954X_CHANNEL_ROOT_PRIO = I2C_TCA954X_CHANNEL_INIT_PRIO + 1. But it's neither possible to do so when defining the device nor in KConfig files. The other solution would be to have the default set to arbitrary chosen value but I suspect this will probably also break for a lot of case. Also the `I2C_TCA954X_CHANNEL_ROOT_PRIO` need to be higher than the i2c bus controller driver init priority the TCA954x is connected to.

I will open an issue to add a compilation error when I2C_TCA954X_CHANNEL_INIT_PRIO <= I2C_TCA954X_CHANNEL_ROOT_PRIO. So that users are aware to set those according to their usage

aaillet commented on drivers/i2c/i2c_tca954x.c:
```diff
@@ -0,0 +1,214 @@
+};
+
+#define INIT_RESET_GPIO(inst, compat)							\
+	COND_CODE_1(DT_NODE_HAS_PROP(DT_INST(inst, compat), reset_gpios),		\
```
forgot to do it when switching to "gpio_dt_spec", just fixed and tested it, thanks

mbolivar-nordic commented on drivers/i2c/i2c_tca954x.c:
```diff
@@ -0,0 +1,208 @@
+
+#define TCA954X_FOREACH_STATUS_OKAY(compat, fn)		\
+	COND_CODE_1(DT_HAS_COMPAT_STATUS_OKAY(compat),		\
+		    (UTIL_CAT(DT_FOREACH_OKAY_INST_,		\
+			      compat)(fn)),			\
```
You are using the generated macros directly here; this is not allowed. These are an implementation detail that should not be used directly from a driver. Please use DT_INST_FOREACH_STATUS_OKAY instead.

aaillet replied:
I think that replacing `TCA954X_FOREACH_STATUS_OKAY` by `DT_FOREACH_STATUS_OKAY` is the cleanest way to do it with DT macros.
Being based on the adc lmp90xxx driver, do you want me to push another PR to fix this in this one too ?

aaillet replied:
After testing, it wont work since `DT_FOREACH_STATUS_OKAY` is declaring `DT_N_S_soc_S_i2c_e6510000_S_pca9548_71_S_mux_i2c_2_REG_IDX_0_VAL_ADDRESS` while my calls for `DT_INST` are searching for `DT_N_INST_DT_N_S_soc_S_i2c_e6510000_S_pca9548_71_ti_tca9548a_REG_IDX_0_VAL_ADDRESS`.

So if `DT_INST_FOREACH_STATUS_OKAY` cannot take a compat as argument and there isn't an equivalent to `DT_INST` that is `UTIL_CAT(DT_N_, DT_DASH(inst, compat))` the way I did it is the only way I see to do it !

aaillet replied:
After some testing, it's not possible to use `DT_INST_FOREACH_STATUS_OKAY` because it wont take compat as argument and we need it as soon as we support multiple compatible.
I tried using `DT_FOREACH_STATUS_OKAY` but it wont work either because `DT_INST` is then searching for a `DT_INST_FOREACH_STATUS_OKAY` generated field.

That's why we use `TCA954X_FOREACH_STATUS_OKAY` to do `DT_INST_FOREACH_STATUS_OKAY` that can take a compat as argument.
I tried multiple ways to do it but can't find any that is working except the one already used.

mbolivar-nordic replied:
You can `#undef DT_DRV_COMPAT` and then define it to something else before using `DT_INST_FOREACH_STATUS_OKAY`. See, for example, drivers/serial/uart_gecko.c.

aaillet replied:
Just tested this :
```
/*
 * TCA9546A: 4 channels
 */
#define TCA9546A_INIT(n) TCA954x_ROOT_DEFINE(n, 4, ti_tca9546a)
#undef DT_DRV_COMPAT
#define DT_DRV_COMPAT ti_tca9546a
DT_INST_FOREACH_STATUS_OKAY(TCA9546A_INIT)

/*
 * TCA9548A: 8 channels
 */
#define TCA9548A_INIT(n) TCA954x_ROOT_DEFINE(n, 8, ti_tca9548a)
#undef DT_DRV_COMPAT
#define DT_DRV_COMPAT ti_tca9548a
DT_INST_FOREACH_STATUS_OKAY(TCA9548A_INIT)
```

It's working, is there a way to do it in a cleaner way or it is what you were thinking of @mbolivar-nordic ? Thanks

aaillet replied:
fixed it

mbolivar-nordic commented on drivers/i2c/Kconfig.tca954x:
```diff
@@ -0,0 +1,19 @@
+menuconfig I2C_TCA954X
+	bool "I2C addressable switch"
+	help
+	  Enable TCA954x serie I2C bus switch
```
```suggestion
	  Enable TCA954x series I2C bus switch
```

aaillet replied:
will fix, thanks

mbolivar-nordic reviewed (changes requested):
Generally the DT bits look fine; thank you. I did see one issue that is a blocker for me, but otherwise I think it's OK.

I noticed a minor typo while I was looking at the Kconfig.

aaillet committed:
driver: i2c: TCA954X: Add support for "reset-gpios"

Add an optional "reset-gpios" phandle-array
field to the driver common yaml.
The reset GPIO channel and pin can be defined
in device tree as a node property.
The driver then deassert the reset signal at mux
initialization if a "reset-gpios" has been specified.

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

aaillet committed:
driver: i2c: TCA954X: Add support for TCA9548A

Add support for the 8 channels TCA9548A I2C mux.
Added a new binding ti,tca9548a binding inheriting
properties from ti,tca954x-base and defining its own compatibles fields.

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

aaillet committed:
driver: i2c: TCA9546A: Rename to TCA954x

Prepare the driver to upcoming support of more mux references.
Rename all TCA9546A related files to TCA954x.
Keep ti,tca9546a and ti,tca9546a-channel compatible
for backward compatibility reasons.
New tca954x-base binding embedding common properties,
tca9546a binding inherits from it and define its own compatibles fields.

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

aaillet commented:
Replaced `TCA954X_FOREACH_STATUS_OKAY` by redefining `DT_DRV_COMPAT` for each device supported.
Also renamed config and data structure in order to support multiple device version on one and only board (for both root and channels structures).
Rebased on top of main.
